### PR TITLE
Integration test for max block size

### PR DESCRIPTION
**Pull request by evan-forbes**
## Description

This PR fixes a quite a few bugs in order to get a new integration test to pass consistently. The test involves 2 validators chewing through 1) 20 ~1Mb PFDs and 2) 100 randomly sized PFDs, along with performing a few simple check on the size of the square for those blocks. This integration test also serves as a great starting point for future tests.

Notably, it also removes the user pre-padding their PFDs by default, which has some significant trade offs that still need to be discussed, likely in a different issue. 

closes: #410, #333, #389, #400, #379

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -158,5 +158,5 @@ replace (
 	github.com/99designs/keyring => github.com/cosmos/keyring v1.1.7-0.20210622111912-ef00f8ac3d76
 	github.com/cosmos/cosmos-sdk => github.com/celestiaorg/cosmos-sdk v1.1.0-sdk-v0.46.0
 	github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
-	github.com/tendermint/tendermint v0.35.4 => github.com/celestiaorg/celestia-core v1.2.1-tm-v0.35.4
+	github.com/tendermint/tendermint v0.35.4 => github.com/celestiaorg/celestia-core v1.2.1-tm-v0.35.4.0.20220518201820-1dc64a37f296
```
Comment (evan-forbes) [inline go.mod:161]: still need to replace this with an actual release after #754

File: x/payment/client/testutil/integration_test.go [hunk 2]
```diff
@@ -82,20 +82,7 @@ func (s *IntegrationTestSuite) TestSubmitWirePayForData() {
...
-		{
-			"valid transaction list of square sizes",
-			[]string{
-				hexNS,
-				hexMsg,
-				fmt.Sprintf("--from=%s", username),
-				fmt.Sprintf("--%s=%s", flags.FlagBroadcastMode, flags.BroadcastBlock),
-				fmt.Sprintf("--%s=%s", flags.FlagFees, sdk.NewCoins(sdk.NewCoin(s.cfg.BondDenom, sdk.NewInt(2))).String()),
-				fmt.Sprintf("--%s=true", flags.FlagSkipConfirmation),
-				fmt.Sprintf("--%s=%s", paycli.FlagSquareSizes, "2,4,8,16,32"),
```
Comment (evan-forbes) [inline x/payment/client/testutil/integration_test.go:98]: this test was essentially the same as the other test, so it was deleted.

File: x/payment/client/testutil/integration_test.go [hunk 3]
```diff
@@ -124,39 +111,43 @@ func (s *IntegrationTestSuite) TestSubmitWirePayForData() {
 			out, err := clitestutil.ExecTestCLICmd(clientCtx, cmd, tc.args)
 			if tc.expectErr {
 				require.Error(err)
-			} else {
```
Comment (evan-forbes) [inline x/payment/client/testutil/integration_test.go:127]: this else statement was superfluous

File: x/payment/types/builder.go [hunk 4]
```diff
@@ -112,14 +137,15 @@ func (k *KeyringSigner) BuildSignedTx(builder sdkclient.TxBuilder, msg sdktypes.
...
+
 	// Generate the bytes to be signed.
 	bytesToSign, err := k.encCfg.TxConfig.SignModeHandler().GetSignBytes(
 		signing.SignMode_SIGN_MODE_DIRECT,
-		authsigning.SignerData{
-			ChainID:       k.chainID,
-			AccountNumber: accountNumber,
-			Sequence:      sequence,
-		},
+		signerData,
```
Comment (evan-forbes) [inline x/payment/types/builder.go:148]: this was another bug where we weren't adding the public key and address to the `SignerData`, so the sign bytes were sometimes being generated incorrectly.

**Review by evan-forbes — COMMENTED** [review]
I'm keeping this PR as a draft for now, as we have yet to merge #754, but it is ready for review.

running the race detector with the integration test that runs 2 validator nodes and 100Mb worth of data causes the test to be super super slow, and seems to timeout no matter how long the timeout is set to, so I'm not sure what we want to do about that. It's probably okay to either not run that test while also running the race detector, as we can still run it during the cli integration test, which is significantly smaller, and should pick up any low handing data races.

**Comment by evan-forbes** [comment]
Marking ready for review while I figure out what to do about the race detector inhibiting the big block integration test to the point of failure

File: app/test/block_size_test.go [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+}
+
+func NewIntegrationTestSuite(cfg cosmosnet.Config) *IntegrationTestSuite {
+	return &IntegrationTestSuite{cfg: cfg}
+}
+
+func (s *IntegrationTestSuite) SetupSuite() {
+	s.T().Log("setting up integration test suite")
+
+	if testing.Short() {
```
Comment (SweeXordious) [inline app/test/block_size_test.go:45]: Is using this better than checking an environment variable?

File: app/test/block_size_test.go [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+		})
+		require.NoError(s.network.WaitForNextBlock())
+	}
+
+}
+
+func TestIntegrationTestSuite(t *testing.T) {
+	cfg := network.DefaultConfig()
+	cfg.EnableTMLogging = false
+	cfg.MinGasPrices = "0uceles"
```
Comment (SweeXordious) [inline app/test/block_size_test.go:165]: shouldn't this follow the new denomination ?

File: x/payment/types/builder.go [hunk 7]
```diff
@@ -64,17 +65,41 @@ func (k *KeyringSigner) QueryAccountNumber(ctx context.Context, conn *grpc.Clien
 	return nil
 }
 
+func (k *KeyringSigner) UpdateAccountFromClient(clientCtx client.Context) error {
+	// clientCtx.AccountRetriever.
```
Comment (SweeXordious) [inline x/payment/types/builder.go:69]: signaling in case you forgot to  delete this comment :D

**Review by SweeXordious — COMMENTED** [review]
I like the `IntegrationTestSuite` way :+1:

File: go.mod (hunk 1 above)
Comment (adlerjohn) [reply go.mod]: 754 doesn't exist in this repo

**Review by adlerjohn — COMMENTED** [review]
CI failing?

**Comment by evan-forbes** [comment]
> CI failing?

yeah, still the same issue we were experiencing before and with other large tests in other repos. Running the race detector on large tests just doesn't work. If we use a single validator and small blocks, it does work (see the cli integration test), but if we use large blocks or two validators it doesn't.

In the past, using a build tag and simply not running the test if the race detector was running, as the build tags are visible at runtime. see https://github.com/golang/go/issues/7007

 I haven't found a solution that I'm happy with atm, we might have to go with the very ugly solution we used before in core

**Review by adlerjohn — COMMENTED** [review]
no comments

File: app/prepare_proposal.go [hunk 8]
```diff
@@ -68,9 +68,8 @@ func (app *App) estimateSquareSize(data *core.Data) uint64 {
 	msgShareEstimate := estimateMsgShares(app.txConfig, data.Txs)
 
 	totalShareEstimate := txShareEstimate + evdShareEstimate + msgShareEstimate
-
-	estimatedSize := types.NextPowerOf2(uint64(math.Sqrt(float64(totalShareEstimate))))
-
+	sr := math.Sqrt(float64(totalShareEstimate))
+	estimatedSize := types.NextHighestPowerOf2(uint64(sr))
```
[the remark this answers is not in the document]
Comment (liamsi) [reply app/prepare_proposal.go:72]: Maybe we should add a comment here too.

File: x/payment/client/testutil/integration_test.go [hunk 9]
```diff
@@ -40,10 +40,6 @@ func NewIntegrationTestSuite(cfg cosmosnet.Config) *IntegrationTestSuite {
 func (s *IntegrationTestSuite) SetupSuite() {
 	s.T().Log("setting up integration test suite")
 
-	if testing.Short() {
-		s.T().Skip("skipping test in unit-tests mode.")
-	}
-
```
Comment (liamsi) [inline x/payment/client/testutil/integration_test.go:46]: Isn't skipping somewhere still necessary?

File: app/split_shares.go [hunk 10]
```diff
@@ -23,10 +23,9 @@ import (
 // block data are only used to avoid dereferening, not because we need the block
 // data to be mutable.
 func SplitShares(txConf client.TxConfig, squareSize uint64, data *core.Data) ([][]byte, *core.Data) {
-	var (
-		processedTxs [][]byte
-		messages     core.Messages
-	)
+	processedTxs := make([][]byte, 0)
+	messages := core.Messages{}
```
[the remark this answers is not in the document]
Comment (liamsi) [reply app/split_shares.go:26]: add comment in code

File: x/payment/types/builder.go (hunk 7 above)
Comment (evan-forbes) [reply x/payment/types/builder.go]: thanks!

File: x/payment/types/payfordata.go [hunk 11]
```diff
@@ -112,16 +114,24 @@ func BuildPayForDataTxFromWireTx(
 // https://github.com/celestiaorg/celestia-specs/blob/master/src/rationale/message_block_layout.md#message-layout-rationale
 func CreateCommitment(k uint64, namespace, message []byte) ([]byte, error) {
 	// add padding to the message if necessary
-	message = padMessage(message)
```
[the remark this answers is not in the document]
Comment (liamsi) [reply x/payment/types/payfordata.go:115]: link to issue?

File: app/test/block_size_test.go (hunk 6 above)
Comment (evan-forbes) [reply app/test/block_size_test.go]: good find! https://github.com/celestiaorg/celestia-app/pull/419/commits/ec45ceb29983e2472a773908f3162d2b1ac3815e

File: x/payment/client/testutil/integration_test.go (hunk 9 above)
Comment (evan-forbes) [reply x/payment/client/testutil/integration_test.go:46]: not for this test, as its small enough for the detector to handle

**Review by liamsi — DISMISSED** [review]
Approving. Any outstanding comments should be addressed in followup issues/PRs.

File: app/test/block_size_test.go (hunk 5 above)
Comment (evan-forbes) [reply app/test/block_size_test.go:45]: imo, yes, as its builtin. Open to other solutions tho

File: app/split_shares.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (evan-forbes) [reply app/split_shares.go]: https://github.com/celestiaorg/celestia-app/pull/419/commits/8928fccb4c60537c56eeff697ba94ff0cc33066c

sorry, didn't see that you already approved before pushing

File: x/payment/types/payfordata.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (liamsi) [reply x/payment/types/payfordata.go:115]: Nope just as you did

**Review by liamsi — APPROVED** [review]

**Pull request merged**